plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/PublishDiagnosticsProcessor.java

```diff
@@ -12,29 +12,61 @@
 
 import com.google.inject.Inject;
 import com.google.inject.Singleton;
+import java.util.List;
+import java.util.Optional;
 import org.eclipse.che.api.languageserver.shared.model.ExtendedPublishDiagnosticsParams;
+import org.eclipse.che.ide.api.data.tree.Node;
 import org.eclipse.che.ide.api.editor.EditorAgent;
 import org.eclipse.che.ide.api.editor.EditorPartPresenter;
 import org.eclipse.che.ide.api.editor.annotation.AnnotationModel;
 import org.eclipse.che.ide.api.editor.editorconfig.TextEditorConfiguration;
 import org.eclipse.che.ide.api.editor.texteditor.TextEditor;
+import org.eclipse.che.ide.part.explorer.project.ProjectExplorerPresenter;
 import org.eclipse.che.ide.resource.Path;
+import org.eclipse.che.ide.resources.tree.ResourceNode;
+import org.eclipse.che.ide.ui.smartTree.Tree;
 import org.eclipse.lsp4j.Diagnostic;
 
 /** @author Anatolii Bazko */
 @Singleton
 public class PublishDiagnosticsProcessor {
 
   private final EditorAgent editorAgent;
+  private final Tree tree;
 
   @Inject
-  public PublishDiagnosticsProcessor(EditorAgent editorAgent) {
+  public PublishDiagnosticsProcessor(
+      EditorAgent editorAgent, ProjectExplorerPresenter projectExplorer) {
     this.editorAgent = editorAgent;
+    this.tree = projectExplorer.getTree();
   }
 
-  public void processDiagnostics(ExtendedPublishDiagnosticsParams diagnosticsMessage) {
-    EditorPartPresenter openedEditor =
-        editorAgent.getOpenedEditor(new Path(diagnosticsMessage.getParams().getUri()));
+  public void processDiagnostics(List<ExtendedPublishDiagnosticsParams> diagnostics) {
+    tree.getNodeStorage()
+        .getAll()
+        .stream()
+        .filter(node -> node instanceof ResourceNode)
+        .forEach(
+            node -> {
+              ((ResourceNode) node).getData().setHasError(false);
+              tree.refresh(node);
+            });
+    diagnostics.forEach(this::processDiagnostics);
+  }
+
+  private void processDiagnostics(ExtendedPublishDiagnosticsParams diagnosticsMessage) {
+    final Path path = Path.valueOf(diagnosticsMessage.getParams().getUri());
+    final Optional<Node> resource =
+        tree.getNodeStorage()
+            .getAll()
+            .stream()
+            .filter(
```

Comment:
Here we also can probably make it more concise and readable making some changes, something similar to this:
```
.filter(ResourceNode.class::isInstance)
.map(ResourceNode.class::cast)
.filter(node -> node.getData().isFile())
.filter(node -> path.equals(node.getData().asFile().getLocation)))
.findAny()
.ifPresent(this::setHasError);
```

What is you opinion?

Reply:
Hi, thanks for reviewing.
Actually this is what I did in the first place but I don't know why, GWT does not compile method reference ResourceNode.class::isInstance.
I get the following error:
The type Class<ResourceNode> does not define isInstance(Node) that is applicable here